server/task_executor_pitr_cutover.go

```diff
@@ -138,3 +137,115 @@ func (exec *PITRCutoverTaskExecutor) pitrCutover(ctx context.Context, task *api.
+		return nil, "", err
+	}
+	log.Debug("Swapping the original and PITR database", zap.String("originalDatabase", database.Name))
+	pitrDatabaseName, pitrOldDatabaseName, err := mysql.SwapPITRDatabase(ctx, conn, database.Name, suffixTs)
+	if err != nil {
+		log.Error("Failed to swap the original and PITR database", zap.String("originalDatabase", database.Name), zap.String("pitrDatabase", pitrDatabaseName), zap.Error(err))
+		return nil, "", fmt.Errorf("failed to swap the original and PITR database, error: %w", err)
+	}
+	log.Debug("Finished swapping the original and PITR database", zap.String("originalDatabase", database.Name), zap.String("pitrDatabase", pitrDatabaseName), zap.String("oldDatabase", pitrOldDatabaseName))
+
+	binlogInfo, err := mysql.GetBinlogInfo(ctx, conn)
+	if err != nil {
+		return nil, "", err
+	}
+	payload := api.BackupPayload{BinlogInfo: binlogInfo}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, "", err
+	}
+
+	log.Debug("Starting new transaction and unlock tables")
+	options := sql.TxOptions{ReadOnly: true}
+	// Beginning a transaction in the same session will implicitly release existing table locks.
+	// ref: https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html, section "Interaction of Table Locking and Transactions".
+	txn, err := conn.BeginTx(ctx, &options)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return txn, string(payloadBytes), nil
+}
+
+func (exec *PITRCutoverTaskExecutor) backupDatabaseAfterPITR(ctx context.Context, conn *sql.Conn, txn *sql.Tx, backupPayload string, driver db.Driver, store *store.Store, database *api.Database, dataDir string, suffixTs int64) error {
+	backupName := fmt.Sprintf("%s-%s-pitr_%d", api.ProjectShortSlug(database.Project), api.EnvSlug(database.Instance.Environment), suffixTs)
+	if err := createBackupDirectory(dataDir, database.ID); err != nil {
+		return fmt.Errorf("failed to create backup directory, error: %w", err)
+	}
+	path := getBackupRelativeFilePath(database.ID, backupName)
+	migrationHistoryVersion, err := getLatestSchemaVersion(ctx, driver, database.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get migration history for database %q, error: %w", database.Name, err)
+	}
+	backupCreate := &api.BackupCreate{
+		CreatorID:               api.SystemBotID,
+		DatabaseID:              database.ID,
+		Name:                    backupName,
+		Type:                    api.BackupTypeAutomaticPITRCutover,
+		MigrationHistoryVersion: migrationHistoryVersion,
+		StorageBackend:          api.BackupStorageBackendLocal,
+		Path:                    path,
+	}
+	backup, err := store.CreateBackup(ctx, backupCreate)
+	if err != nil {
+		if common.ErrorCode(err) == common.Conflict {
+			return fmt.Errorf("backup name %q already exists", backupName)
+		}
+		return fmt.Errorf("failed to create backup %q, error: %w", backupName, err)
+	}
+
+	backupPatch := api.BackupPatch{
+		ID:        backup.ID,
+		Status:    string(api.BackupStatusDone),
```

Comment:
I prefer to set `Status: string(api.BackupStatusFailed)` first because there are a lot of errors we may encounter.

Reply:
The current occurrence of `backupPatch` is misleading. I moved it to the usage side, PTAL again.